src/utilities/controllers/controller_io.hpp

```diff
@@ -1,100 +1,117 @@
+    int status = 0;    // (1) -->  Status flag set as follows: 0 if this is the first call, 1 for all
+                       // subsequent time steps, -1 if this is the final call at the end of the
+                       // simulation (-)
+    double time = 0.;  // (2) -->  Current time (sec) [t in single precision]
+    double dt = 0.;    // (3) -->  Communication interval (sec)
+    double pitch_blade1 = 0.;              // (4) -->  Blade 1 pitch angle (rad)
+    double pitch_angle_actuator_req = 0.;  // (10)
+    double generator_speed = 0.;           // (20) Generator speed (RPM)
+    double horizontal_wind_speed = 0.;     // (27) Horizontal hub-heigh wind speed (m/s)
+    double pitch_blade2 = 0.;              // (33) Blade 2 pitch
+    double pitch_blade3 = 0.;              // (34) Blade 3 pitch
+    int generator_contactor_status = 0;    // (35) Generator contactor status: 1=main (high speed)
+                                           // variable-speed generator
+    int shaft_brake_status = 0;            // (36) Shaft brake status: 0=off
+    double demanded_yaw_actuator_torque = 0.;  // (41) Demanded yaw actuator torque
+    double pitch_command_1 = 0.;  // (42) Use the command angles of all blades if using individual
+                                  // pitch
+    double pitch_command_2 = 0.;  // (43) Use the command angles of all blades if using individual
+                                  // pitch
+    double pitch_command_3 = 0.;  // (44) Use the command angles of all blades if using individual
+                                  // pitch
+    double pitch_command_collective = 0.;      // (45) Use the command angle of blade 1 if using
+                                               // collective pitch
+    double demanded_pitch_rate = 0.;           // (46) Demanded pitch rate (Collective pitch)
+    double demanded_generator_torque = 0.;     // (47) Demanded generator torque
+    double demanded_nacelle_yaw_rate = 0.;     // (48) Demanded nacelle yaw rate
+    size_t message_array_size = 0U;            // (49) avcMSG array size
+    size_t infile_array_size = 0U;             // (50) avcINFILE array size
+    size_t outname_array_size = 0U;            // (51) avcOUTNAME array size
+    int pitch_override = 0;                    // (55) Pitch override: 0=yes
+    int torque_override = 0;                   // (56) Torque override: 0=yes
+    size_t n_blades = 0U;                      // (61) Number of blades
+    size_t n_log_variables = 0U;               // (65) Number of variables returned for logging
+    double generator_startup_resistance = 0.;  // (72) Generator start-up resistance
+    int loads_request = 0;                     // (79) Request for loads: 0=none
+    int variable_slip_status = 0;              // (80) Variable slip current status
+    int variable_slip_demand = 0;              // (81) Variable slip current demand
+
+    void CopyToSwapArray(std::array<float, kSwapArraySize>& swap_array) const {
```

Comment:
These copy to/from swap array functions look like they're going to be difficult to maintain with the same variable needing to be in three different locations and the correct array indices added in two places. I thought you had been talking about pushing the swap array into Device memory and having the constraints access it more directly.

Reply:
We'll have to add the mapping as we add new features that we support, but that should be pretty "write once, change never."  A test for each entry should be pretty easy to write, thus proving correctness.  I agree that it's verbose, but I think it's worth it to remove the need for a reinterpret cast of raw memory allocated by another codebase.  

I am still thinking about doing something along those lines, but those changes would be pretty involved, so I'm leaving it for a future step.  The goal here was to be otherwise transparent to our current usage.